Xamarin.Forms.Platform.iOS/CollectionView/ObservableGroupedSource.cs

```diff
@@ -127,45 +129,53 @@ void ResetGroupTracking()
 			}
 		}
 
-		void CollectionChanged(object sender, NotifyCollectionChangedEventArgs args)
+		async void CollectionChanged(object sender, NotifyCollectionChangedEventArgs args)
 		{
 			if (Device.IsInvokeRequired)
 			{
-				Device.BeginInvokeOnMainThread(() => CollectionChanged(args));
+				Device.BeginInvokeOnMainThread(async () => await CollectionChanged(args));
 			}
 			else
 			{
-				CollectionChanged(args);
+				await CollectionChanged(args);
 			}
 		}
 
-		void CollectionChanged(NotifyCollectionChangedEventArgs args)
+		async Task CollectionChanged(NotifyCollectionChangedEventArgs args)
 		{
 			switch (args.Action)
 			{
 				case NotifyCollectionChangedAction.Add:
-					Add(args);
+					await Add(args);
 					break;
 				case NotifyCollectionChangedAction.Remove:
-					Remove(args);
+					await Remove(args);
 					break;
 				case NotifyCollectionChangedAction.Replace:
-					Replace(args);
+					await Replace(args);
 					break;
 				case NotifyCollectionChangedAction.Move:
 					Move(args);
 					break;
 				case NotifyCollectionChangedAction.Reset:
-					Reload();
+					await Reload();
 					break;
 				default:
 					throw new ArgumentOutOfRangeException();
 			}
 		}
 
-		void Reload()
+		async Task Reload()
 		{
 			ResetGroupTracking();
+
+			while (_batchUpdating)
+			{
+				// We don't want to reload while the UICollectionView is animating changes; if we do, things get
+				// broken or out-of-order. So just hang tight until it's finished.
+				await Task.Delay(16);
```

Comment:
could this be achieved using Semaphoreslim?
https://github.com/xamarin/Xamarin.Forms/blob/52a2581dfe4bd52f8bcabfb714c63e784735fee3/Xamarin.Forms.Platform.Android/ImageCache.cs#L38

Reply:
Or maybe a TaskCompleteSource?

Reply:
I like it, much cleaner. Done.